### DISCUSSION TITLE
addon.json 2.0

### DISCUSSION THREAD

**Discussion by DNin01** [ScratchAddons/ScratchAddons#6528] [Discussion]
I was wondering, what if the addon manifest got a major, non-backwards-compatible update?

#### New and improved `if`

Using `if` inside styles and settings allows them to be active/shown only when certain addons or settings are enabled, but it's pretty primitive right now, everything is a logical OR and you can't show something when an addon is <i>dis</i>abled. There are some changes we can make to `if` to make it more powerful, and some things we can fix:
- #6522
- #6523
- #5692
- #6655
- #5272

One very enticing idea would be to support the syntax for conditions in JavaScript here. An example from https://github.com/ScratchAddons/ScratchAddons/issues/6522#issuecomment-1692699823:
```json
{
  "if": "addonEnabled('example') || self.settings('one') || self.settings('two')"
}
```

#### One line to activate dynamic everything

Instead of declaring `dynamicDisable`, `dynamicEnable`, and `updateUserstylesOnSettingsChange` over and over again just to get their corresponding features to work, we could replace these with a new, more compact key. I'm not sure what to call it, but I'll just say `dynamic`:
```json
"dynamic": true
```
That would be the equivalent of setting all three of the aforementioned properties to true. And if you wanted to just make an addon dynamically enablable and have the userstyles adapt to setting changes, use an array:
```json
"dynamic": ["enable", "configure"]
```
(...Have the userstyles... with `"configure"`... ah, making that intuitive might take some thinking. Alternatively, we can try #3315.)

#### They're not really "user" scripts

Currently, references to JS and CSS files are declared inside the keys `userscripts` and `userstyles` respectively, but in reality, [userscripts](https://en.wikipedia.org/wiki/Userscript) are actually a pretty different topic and might be interpreted as something the user creates and an extension runs. I know this isn't really necessary, but addon manifest 2.0 would be a good time to change this if someone really wants to. In that case, these keys could be changed to `scripts` and `styles`, or even just `js` and `css`.

### Any other ideas or related issues?

[Find issues](https://github.com/ScratchAddons/ScratchAddons/issues?q=is%3Aissue+is%3Aopen+label%3A%22scope%3A+addon.json%22+label%3A%22type%3A+enhancement%22)

**Comment by Samq64** [comment]
If this ever happens I think some key names should be shortened since I don't think they need to be very descriptive.

**Reply by WorldLanguages** [reply]
I would like to see a list of names (not addon IDs!) that are confusing. They can also be addon.* API terms.

**Comment by WorldLanguages** [comment]
Let's first consider #3315, which says `dynamicEnable`, `dynamicDisable`, and others, should be togglable for each userstyle.

We end up with these ways to classify userstyles:
- Userstyles that aren't dynamic at all. They need a reload to be injected, and a reload to be removed. The condition can be nothing (just "this addon enabled"), or it could be and `"if"` condition in the userstyle, which will only be evaluated when the page is loading.
- Userstyles that can be dynamically enabled, but they cannot be removed after that. If there's an `"if"` condition, no need to evaluate it after the userstyle has been dynamically injected into all tabs that are currently open.
- Userstyles that are perfectly dynamic and are checked against settings each time there's any change to user settings.

Regarding `"if"` falling short of all logic expressions, you said:

> Using `if` inside styles and settings allows them to be active/shown only when certain addons or settings are enabled, but it's pretty primitive right now, everything is a logical OR and you can't show something when an addon is disabled. There are some changes we can make to `if` to make it more powerful, and some things we can fix:

At that point, I would suggest something way more flexible: custom JS logic on whether a userstyle should be injected into a page or not.

This idea is quite simple, it's just a small JS module that includes a function that is ran as if it was an `"if"` condition, but you can express any logical expressions with it, of course.

```js
export default function ({ addons, url }) {
  if (addons["editor-dark-mode"].enabled === false && addons["dark-www"].settings["blah"] === "bleh") return true;
  else if (addons["scratchr2"].enabled) return true;
  else return false;
}
```

Of course, the current static `"if"` syntax can be preserved too.

**Reply by DNin01** [reply]
I think it would be nice to avoid using JS modules when possible, although on paper they are pretty cool. Maybe we could expand the feature set of `if` _and_ support using modules when necessary.

**Reply by WorldLanguages** [reply]
I'm not sure if any other attempt to add expressive logic to "if" would be superior. We can't allow inline JS inside a string, unless we want to build our own little parser for this use case. We could also add some way to specify AND/OR operators through JSON, but then it gets hard to read.

**Reply by DNin01** [reply]
Alternatively, all addon metadata (including conditions) could be declared by scripts? So addon.json would become addon.js, kinda like TurboWarp extensions? That way, it could also evaluate conditions.

Nah, that's probably be less efficient.

**Reply by WorldLanguages** [reply]
I've seen that idea multiple times on Discord. There's some benefits to having statically analyzable addon manifests, for example it makes it easy to auto-update https://scratchaddons.com/addons

**Reply by WorldLanguages** [reply]
But yeah, no need to make the userscripts, userstyles, customCssVariables, and some other very technical properties statically-analyzable. We could support declaring those as scripts in separate files.

**Reply by DNin01** [reply]
> But yeah, no need to make the userscripts, userstyles, customCssVariables, and some other very technical properties statically-analyzable. We could support declaring those as scripts in separate files.

Reminds me of how developers typically separate styles and scripts into separate files instead of self-containing them in the HTML source, to keep things organized. I see potential here.

**Reply by WorldLanguages** [reply]
Well, we already do that... But our addon.json files are somewhat user-readable and we want to add more specific logic on when certain styles should be active.

**Reply by DNin01** [reply]
We could also use a build step to get around the `eval()` problem.

**Reply by WorldLanguages** [reply]
Including JS code outside of .js also has other drawbacks: ESLint, prettier, intellisense, and others, don't notice that it's code.

It would be a better idea (but unsure if a good idea) to have a manifest.js that is directly used by the extension, and a built static addon.json file that is kept for other use cases that require a raw json file for simplicity.

**Comment by mybearworld** [comment]
This could also standardize whether to use `"url"` or `"link"`, as that's currently incosistent.

**Reply by mxmou** [reply]
I think the current names make sense: `link` creates an actual link.
